File: Data Structures/Segment Tree/Segment Tree.cpp
// access is done in 0 indexing
// calculations are done in 1 indexing
template <class T>
class SegmentTree {
 private:
  int n;
  T neutral;
  vector<T> t;

  // the main operation of the tree
  T operation(T l, T r) { return l * r; }

  void build() {
    for (int i = n - 1; i > 0; --i)
      t[i] = operation(t[i << 1], t[i << 1 | 1]);
  }

 public:
  // Initialize empty
  SegmentTree(int n, T neutral) {
    this->n = n;
    this->neutral = neutral;
    t.assign(n * 2, neutral);
  }

  // Initialize from array
  SegmentTree(vector<T> arr, T neutral) {
    this->n = arr.size();
    this->neutral = neutral;
    t.assign(n * 2, neutral);
    for (int i = 0; i < n; i++) {
      t[n + i] = arr[i];
    }
    build();
  }

  // set value at position i
  void modify(int i, T value) {
    for (t[i += n] = value; i >>= 1;)
      t[i] = operation(t[i << 1], t[i << 1 | 1]);
  }

  // query on interval [l, r)
  T query(int l, int r) {
    T resl = neutral, resr = neutral;
    for (l += n, r += n; l < r; l >>= 1, r >>= 1) {
      if (l & 1)
        resl = operation(resl, t[l++]);
      if (r & 1)
        resr = operation(t[--r], resr);
    }
    return operation(resl, resr);
  }

  void clear() {
    t.clear();
    n = 0;
  }
};

File: Data Structures/Segment Tree/Segment Tree.py
class SegmentTree:
    def __init__(self, neutral, arr=None, n=None):
        self.neutral = neutral
        self.n = n if n else len(arr)
        self.t = [self.neutral] * (self.n * 2)
        if arr:
            for i in range(self.n):
                self.t[self.n + i] = arr[i]
            self.build()

    # the main operation of the tree
    def operation(self, l, r):
        return l * r

    def build(self):
        i = self.n - 1
        while i > 0:
            self.t[i] = self.operation(self.t[i << 1], self.t[i << 1 | 1])
            i -= 1

    # set value at position i
    def modify(self, i, value):
        i += self.n
        self.t[i] = value
        i >>= 1
        while i > 0:
            self.t[i] = self.operation(self.t[i << 1], self.t[i << 1 | 1])
            i >>= 1

    # query on interval [l, r)
    def query(self, l, r):
        resl, resr = self.neutral, self.neutral
        l += self.n
        r += self.n
        while l < r:
            if l & 1:
                resl = self.operation(resl, self.t[l])
                l += 1
            if r & 1:
                r -= 1
                resr = self.operation(self.t[r], resr)
            l >>= 1
            r >>= 1
        return self.operation(resl, resr)
